### PR TITLE
Fix major bug in v0.3.1: negative counts

### LINKED ISSUE #306: negative entries in filtered count matrix?
Hi,

I've been using this amazing tool successfully for more than three years now. two days ago I updated (`git pull && pip install`) to the most recent version (0.3.1; commit 3a4dc8), and I'm getting weird results: there are negative entries in the filtered h5 file, and the HTML report looks weird too. This happened in all the datasets I've run, so I tried again with [Brain single-nuc](https://cf.10xgenomics.com/samples/cell-exp/7.0.1/SC3pv3_GEX_Human_PBMC/SC3pv3_GEX_Human_PBMC_raw_feature_bc_matrix.h5) and [PBMC single-cell](https://cf.10xgenomics.com/samples/cell-exp/7.0.0/5k_mouse_brain_CNIK_3pv3/5k_mouse_brain_CNIK_3pv3_raw_feature_bc_matrix.h5) datasets from 10X, like so
```
cellbender remove-background --input Brain_10X/5k_mouse_brain_CNIK_3pv3_raw_feature_bc_matrix.h5 --output Brain_10X/Brain_10X.h5 --fpr 0.01 --epochs 150 --cuda --expected-cells 5000 --total-droplets-included 50000
```
but I'm getting the same result: there are about 0.1 and 0.3% negative entries in the filtered count matrix (loaded into R using [this](https://github.com/samuel-marsh/scCustomize/blob/58529b4dc870860c64d5c96eb078f353d7c4ca2b/R/Read_%26_Write_Data.R#L1054) helper function).

Here's a plot comparing total counts pre- and post cellbender for Brain:
mouse_brain_raw_vs_filt
(adding up absolute counts results in post-cellbender results being much larger than pre-cellbender)

the HTML report also contains negative entries for `n_cellbender` or `fraction_remaining`:
[Brain_10X_report.zip](https://github.com/broadinstitute/CellBender/files/13305218/Brain_10X_report.zip)

(worryingly, Seurat doesn't even complain about negative counts and fails with cryptic errors only if you use `vars.to.regress` or `TransferData`)

Am I missing something? this happens with the default `mckp` estimator but also with `map` and `mean`, and without or with `PRmu` posterior regularization. 

thanks for your help!

### LINKED ISSUE #348: HTML output fails due to lxml change
Jupyter's `nbconvert`, which we use to generate an HTML report from jupyter notebook template, recently runs into 
```pytb
Traceback (most recent call last):
  File "/home/sfleming/miniforge3/envs/cellbender/bin/jupyter-nbconvert", line 5, in <module>
    from nbconvert.nbconvertapp import main
  File "/home/sfleming/miniforge3/envs/cellbender/lib/python3.7/site-packages/nbconvert/__init__.py", line 5, in <module>
    from .exporters import *
  File "/home/sfleming/miniforge3/envs/cellbender/lib/python3.7/site-packages/nbconvert/exporters/__init__.py", line 1, in <module>
    from .asciidoc import ASCIIDocExporter
  File "/home/sfleming/miniforge3/envs/cellbender/lib/python3.7/site-packages/nbconvert/exporters/asciidoc.py", line 9, in <module>
    from .templateexporter import TemplateExporter
  File "/home/sfleming/miniforge3/envs/cellbender/lib/python3.7/site-packages/nbconvert/exporters/templateexporter.py", line 25, in <module>
    from lxml.html.clean import clean_html
  File "/home/sfleming/miniforge3/envs/cellbender/lib/python3.7/site-packages/lxml/html/clean.py", line 21, in <module>
    ) from None
ImportError: lxml.html.clean module is now a separate project lxml_html_clean.
Install lxml[html_clean] or lxml_html_clean directly.
```

https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1964938.html

Putting `lxml_html_clean` into requirements.txt resolves the issue.

### PR DESCRIPTION
**Pull request by sjfleming**
Closes #306 
Closes #348 

Something about the update from v0.3.0 to v0.3.1 induced a bug (I'm not yet sure of the scope, but potentially affecting all datasets) where negative counts could be produced in the output.  It is assumed that these outputs are totally compromised.  v0.3.1 has already been redacted, but the master branch is still compromised.

I think the bug probably lies in the updated posterior generation code, which was the target of memory-saving updates in v0.3.1.

Expand tests covering that code.

- unit tests for posterior generation
- integration tests ensuring no negative count matrix entries in the output (not explicitly tested currently)

### CODE REVIEW

**Comment by sjfleming** [comment]
After running simulated data through v0.3.0 and the master branch, I see that the posterior generated in each case is identical.

So despite what I initially thought, it seems like the bug must have been introduced in `estimation.py`, which is quite well-covered with tests...

(It is not a CPU / GPU issue, as a run without `--cuda` basically agrees with a GPU run.)

**Comment by sjfleming** [comment]
It came down to a small typo / miscalculation about what the maximum possible number of noise counts can be.  I had put `np.uint8` in `estimation.py` line 241 [here](https://github.com/broadinstitute/CellBender/commit/3a4dc8de7157f78e71c1689a536392a3e02317e8), but `np.uint8` is not large enough.  This was probably causing integer overflows that affected the noise estimation process in unexpected ways (which admittedly I have not completely understood yet).

Changing to `np.uint8` seems to fix the problem.

**Before**
image
And there are negative count matrix entries in the output.

**After**
image
And there are no longer negative count matrix entries in the output.

File: cellbender/remove_background/tests/test_estimation.py [hunk 1]
```diff
@@ -92,11 +93,12 @@ def test_mean_massive_m(log_prob_coo):
     new_shape = (coo.shape[0] + greater_than_max_int32, coo.shape[1])
     new_coo = sp.coo_matrix((coo.data, (new_row, coo.col)),
                             shape=new_shape)
+    print(new_coo)
```
Comment (mbabadi) [inline cellbender/remove_background/tests/test_estimation.py:96]: Either remove the print statement, or convert it to something more informative-- e.g. old matrix shape = ..., new matrix shape = ...

**Review by mbabadi — CHANGES_REQUESTED** [review]
Nice job expanding the tests!  Just a small comment to address.  Also, while you're at it, you may want to search the rest of the codebase for uint8 and other explicit dtypes and refactor accordingly.

**Comment by sjfleming** [comment]
Alright @mbabadi , what do you think now?  I added a warning to the README, and I added a command-line argument to enable people to re-run their old checkpointed v0.3.1 runs to produce a valid output.

**Review by mbabadi — APPROVED** [review]
Looks good!

**Pull request merged**